src/test/groovy/org/prebid/server/functional/tests/privacy/GppAuctionSpec.groovy

```diff
@@ -0,0 +1,180 @@
+        def response = defaultPbsService.sendAuctionRequest(bidRequest)
+
+        then: "Bid response should contain warning"
+        assert response.ext?.warnings[ErrorType.PREBID]?.collect { it.code } == [999]
+        assert response.ext?.warnings[ErrorType.PREBID]?.collect { it.message } ==
+                ["GPP scope does not match TCF2 scope"]
+    }
+
+    def "PBS should emit warnings when regs.gpp_sid not contains 2 and regs.gdpr isn't 0"() {
+        given: "Default bid request with gppSid and gdpr"
+        def bidRequest = BidRequest.defaultBidRequest.tap {
+            regs.gdpr = 1
+            regs.gppSid = [PBSUtils.getRandomNumberWithExcept(2)]
+        }
+
+        when: "PBS processes auction request"
+        def response = defaultPbsService.sendAuctionRequest(bidRequest)
+
+        then: "Bid response should contain warning"
+        assert response.ext?.warnings[ErrorType.PREBID]?.collect { it.code } == [999]
+        assert response.ext?.warnings[ErrorType.PREBID]?.collect { it.message } ==
+                ["GPP scope does not match TCF2 scope"]
+    }
+
+    def "PBS should emit warning when GPP string invalid"() {
+        given: "Default bid request with invalid gpp"
+        def bidRequest = BidRequest.defaultBidRequest.tap {
+            regs.gpp = "Invalid_GPP_Consent_String"
+        }
+
+        when: "PBS processes auction request"
+        def response = defaultPbsService.sendAuctionRequest(bidRequest)
+
+        then: "Bid response should contain warning"
+        assert response.ext?.warnings[ErrorType.PREBID]?.collect { it.code } == [999]
+        assert response.ext?.warnings[ErrorType.PREBID]?.collect { it.message } ==
+                ["GPP string invalid: Undecodable FibonacciIntegerRange '101111'"]
+    }
+
+    def "PBS should copy regs.gpp to user.consent when gppSid contains 2, gpp is TCF2-EU and user.consent isn't specified"() {
+        given: "Default bid request with gpp and gppSid"
+        def gppConsent = new GppConsent().setFieldValue()
+        def bidRequest = BidRequest.defaultBidRequest.tap {
+            regs.gppSid = [2]
+            regs.gpp = gppConsent
+            user = new User().tap {
+                consent = null
+            }
+        }
+
+        when: "PBS processes auction request"
+        defaultPbsService.sendAuctionRequest(bidRequest)
+
+        then: "Bidder request should contain user.consent from regs.gpp"
+        def bidderRequest = bidder.getBidderRequest(bidRequest.id)
+        assert bidderRequest.user.consent == gppConsent as String
+        assert bidderRequest.regs.gpp == gppConsent as String
+    }
+
+    def "PBS should emit warning when gppSid contains 2, gpp is TCF2-EU and regs.gpp and user.consent are different"() {
+        given: "Default bid request with gpp and gppSid"
+        def bidRequest = BidRequest.defaultBidRequest.tap {
+            regs.gppSid = [2]
+            regs.gpp = new GppConsent().setFieldValue()
```

Comment:
Implicit reliance on sensitive defaults. Please don't do it that way.

Reply:
We'll rewrite this after discussion.